rust/src/database/verkle/state.rs

```diff
@@ -0,0 +1,589 @@
+// Copyright (c) 2025 Sonic Operations Ltd
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at soniclabs.com/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+use std::mem::MaybeUninit;
+
+use sha3::{Digest, Keccak256};
+
+use crate::{
+    CarmenState,
+    database::verkle::{
+        embedding::{
+            code, get_basic_data_key, get_code_chunk_key, get_code_hash_key, get_storage_key,
+        },
+        variants::SimpleInMemoryVerkleTrie,
+        verkle_trie::VerkleTrie,
+    },
+    error::Error,
+    types::{Address, Hash, Key, Nonce, U256, Update, Value},
+};
+
+const EMPTY_CODE_HASH: Hash = [
+    197, 210, 70, 1, 134, 247, 35, 60, 146, 126, 125, 178, 220, 199, 3, 192, 229, 0, 182, 83, 202,
+    130, 39, 59, 123, 250, 216, 4, 93, 133, 164, 112,
+];
+
+/// An implementation of [`CarmenState`] that uses a Verkle trie as the underlying data structure.
+pub struct VerkleTrieState<T: VerkleTrie> {
+    trie: T,
+}
+
+impl VerkleTrieState<SimpleInMemoryVerkleTrie> {
+    pub fn new() -> Self {
+        let trie = SimpleInMemoryVerkleTrie::new();
+        Self { trie }
+    }
+}
+
+impl<T: VerkleTrie> CarmenState for VerkleTrieState<T> {
+    fn account_exists(&self, addr: &Address) -> Result<bool, Error> {
+        Ok(self.get_code_hash(addr)? != Hash::default())
```

Comment:
I have not looked into the exact semantics, but why not check the nonce instead? EOAs may not have code.